### DISCUSSION TITLE
WiFi drivers - from 0 to hero

### DISCUSSION THREAD

**Discussion by tiagogalvao** [morrownr/8814au#51] [General]
Hello morrownr!
Just made this thread to try to learn something and try to provide help with some wifi drivers.
I have noticed already that is no easy task, so if you want to point me to some directions I appreciate!
Some background: Im a C# developer, but my experience is more "comercial" than anything. I can easily read code in other languages and also adapt as needed.

Drivers are something that I have never touched, so it may take me some time to understand some main concepts and/or be good at it.

Thank you so much for such opportunity! <3 

Edit: I do have the following chips with me
- 8814au (Alfa AWU1900)
- TP LINK TL-WN722N
- TP LINK TL-WN821N

**Comment by morrownr** [comment]
Hi @tiagogalvao 

This is a good thing. I think we will enjoy it and maybe some others will join in.

A little background on me: I learned programming on an IBM mainframe. My first programming language was FORTRAN. We didn't have interactive terminals yet so we had to use punch cards so the programs ran as a batch. I am not a career programmer as life had other plans for me. My primary career was in aviation with a secondary career in education. I am retired. My history of operating systems on small computers looks like this: CP/M, IBM PC DOS, DR DOS, OS/2 and Linux. I have self-taught numerous programming languages over the years. I had only worked on 2 device drivers before starting this project. 

My experience with this project has taught me just how complicated wireless programming can be. A lot of seemingly little issues can turn into a deep dive into the abyss. I would like to spend more time improving this driver but my goal was to improve usb wifi adapter satisfaction for Linux users so I have to be careful to spend my time wisely and on plan. So far I would rate the following repos, in order, as having the greatest impact toward the goal:

https://github.com/morrownr/USB-WiFi

The reason for that repo is that the single biggest problem with usb wifi satisfaction among Linux users was the lack of good information. It was not a new problem but had never been properly corrected. The simple truth is that 90% of Linux users that have a use for usb wifi, need to stick with adapters that are supported with in-kernel drivers as that eliminates a lot of problems. USB-WiFi has grown to the point that I'm seeing over 1,000 hits per week and it continues to grow. How fast it has grown has surprised me. How few problems have been reported has also surprised me .

Another thing that has really helped is getting good drivers in a dependable package that has now been tested out there for all to use. The following drivers are the best Realtek drivers I've ever seen: (And I still have features that I need to document.)

https://github.com/morrownr/8812au-20210629
https://github.com/morrownr/8821au-20210708
https://github.com/morrownr/88x2bu-20210702

Unfortuantely I do not have recent source for the 8814au. The source in use here is from 2019 and it was not a good driver. Realtek has not maintained this driver very well. It is fragile and full of bugs.

Where do we start? I need to know which Linux world you live in. I live in the Debian world but test on some non-Debian systems.

Here is what I recommend for lesson one to give you a big picture:

After you establish dev directory for this adventure, clone this driver in it and also the one above for the 8812au-20210629. Install `meld`

$ sudo apt install meld.

Become familiar with meld and compare the 8814au with the 8812au. Come back with your thoughts and let's see where it goes.

Regards

**Comment by tiagogalvao** [comment]
Hello hello!
Again, thank you so much!
I am reviewing a lot of stuff as I have a lot to cover. I may be slow to reply but I don't plan to make it a waste of your time.
I will post again soon, I just wanted to tell that I haven't vanished or anything lol
